src/coreclr/src/jit/gentree.h

```diff
@@ -4501,12 +4581,65 @@ struct GenTreeIntrinsic : public GenTreeOp
 
 struct GenTreeJitIntrinsic : public GenTreeOp
 {
-    var_types gtSIMDBaseType; // SIMD vector base type
-    unsigned  gtSIMDSize;     // SIMD vector size in bytes, use 0 for scalar intrinsics
+private:
+    ClassLayout* m_layout;
+
+    union {
+        var_types      gtOtherBaseType; // For AVX2 Gather* intrinsics
+        regNumberSmall gtOtherReg;      // For intrinsics that return 2 registers
+    };
+
+public:
+    var_types     gtSIMDBaseType; // SIMD vector base type
+    unsigned char gtSIMDSize;     // SIMD vector size in bytes, use 0 for scalar intrinsics
+
+#if defined(FEATURE_SIMD)
+    union {
+        SIMDIntrinsicID gtSIMDIntrinsicID; // operation Id
+        NamedIntrinsic  gtHWIntrinsicId;
+    };
+#else
+    NamedIntrinsic gtHWIntrinsicId;
+#endif
+
+    ClassLayout* GetLayout() const
+    {
+        return m_layout;
+    }
+
+    void SetLayout(ClassLayout* layout)
+    {
+        assert(layout != nullptr);
+        m_layout = layout;
+    }
+
+    regNumber GetOtherReg() const
+    {
+        return (regNumber)gtOtherReg;
+    }
+    void SetOtherReg(regNumber reg)
+    {
+        gtOtherReg = (regNumberSmall)reg;
+        assert(gtOtherReg == reg);
+    }
+
+    var_types GetOtherBaseType() const
+    {
+        return gtOtherBaseType;
```

Comment:
Do we need to check that `gtHWIntrinsicId == NI_AVX2_GatherVector128 || gtHWIntrinsicId == NI_AVX2_GatherVector256` or is it legal to call `GetOtherBaseType` not only for AVX2 Gather* intrinsic?

Reply:
If we add that assert, we'd have to add a check to the copy method, which I don't think is productive.

Reply:
Nit: I like that the new fields are private, but I find it confusing that we access `gtOtherReg` using this method, maybe we can make them separate fields? It won't change the node size because we allocate them as large nodes. Feel free to ignore that comment.